### ISSUE TITLE
Some remarks on version 3.

### ISSUE THREAD

**Issue by dubinsky** [xmlresolver/xmlresolver#44]
First of all, congratulations and thanks for the major cleanup!

Some remarks:

- [x] 1. [mkdir()](https://github.com/xmlresolver/xmlresolver/blob/main/src/main/java/org/xmlresolver/cache/ResourceCache.java#L161) in `ResourceCache` when creating the cache directory should really be `mkdirs()`: with the `cache` directory inside the `.xmlresolver.org` one, one-step `mkdir()` is guaranteed to fail where the `.xmlresolver.org` directory does not exist. (Cache subdirectories are created one level at a time, so `mkdir()` [works](https://github.com/xmlresolver/xmlresolver/blob/main/src/main/java/org/xmlresolver/cache/ResourceCache.java#L564) there.)
  - confirmed fixed
- [x] 2. Version 3.0.1 caches (URIs rewritten to local) `file:` URIs, not just the ["legacy V2"](#43 ); version 1.x did not. Current 3.0.2-SNAPSHOT continues to do the same; this is probably caused by the failure to [write](https://github.com/xmlresolver/xmlresolver/blob/main/src/main/java/org/xmlresolver/cache/ResourceCache.java#L224) the `control.xml` file, which in turn seems to be caused by the underlying parser (in my case, `SAXParserFactory.newInstance()` finds Saxon 6.5.2's aelfred) throwing an `EOFException` (and [not](https://github.com/xmlresolver/xmlresolver/blob/main/src/main/java/org/xmlresolver/cache/ResourceCache.java#L212) the `FileNotFoundException`) when trying to parse the non-existent `control.xml` file.
  - confirmed fixed
- [x] 3. With the cache disabled (because of the problem 1 above or intentionally, e.g. by setting `CACHE_UNDER_HOME` to `false`), such URIs fail to resolve; see [below](#issuecomment-878478009).
  - works as designed: the parser is going to turn around and arrive at exactly the same result
- [x] 4. `ValidatingXmlLoader` uses its own private `XMLResolverConfiguration` [instance](https://github.com/xmlresolver/xmlresolver/blob/main/src/main/java/org/xmlresolver/loaders/ValidatingXmlLoader.java#L47) with the cache disabled, so the URIs mentioned above fail to resolve when `ValidatingXmlLoader` is used.
  - confirmed fixed
- [x] 5. For `ValidatingXmlLoader` to work at all, dependency `'org.relaxng:jing:20181222'` has to be available; this probably should be documented...
  - documented 
- [x] 6. Resolver is now sensitive to the "uriStartString" ending with `/`; version 1.x was not.
  - works as designed 
- [x] 7. Different implementations of slf4j can be used in different environments (e.g., on Google Cloud Run, logback is likely to be used; inside a Gradle plugin Gradle-supplied implementation is used, etc.). xmlresolver brings in 'org.slf4j:slf4j-simple', which needs to be excluded in the Gradle build files of the project using xmlresolver. In my humble opinion, xmlresolver should not have (production) dependency on any slf4j implementations... (And: all `implementation` dependencies are unnecessarily listed [also](https://github.com/xmlresolver/xmlresolver/blob/main/build.gradle#L33) as `runtimeOnly`; 'org.slf4j:slf4j-api' may as well be listed as `compileOnly` and not as `implementation` dependency.)
  - confirmed fixed

**Comment by ndw** [comment]
Thanks for the observations. I'll fix these today, though I'll have to see if I can recreate the problem where file: URIs are cached. I thought I'd fixed that.

**Comment by ndw** [comment]
Actually, on further reflection, I'm confused by remarks 2, 3, and 4. Perhaps we're discussing two different problems. With `file:` URIs in the cache, you could get into the situation where the cached versions were being used rather than the "original" `file:` URIs on disk (which might have been changed since cached versions were created). But with the cache disabled, this problem can't occur. And the `ValidatingXmlLoader` is only concerned with validating the catalog XML file, it has no impact on URI resolution. Can you clarify the issue you're describing?

**Comment by ndw** [comment]
Can you explain what you mean by `Resolver is now sensitive to the "uriStartString" ending with /`? AFAICT, the match has always been against the start string, taken literally, but perhaps I'm misunderstanding your comment.

**Comment by dubinsky** [comment]
> Can you explain what you mean by `Resolver is now sensitive to the "uriStartString" ending with /`? AFAICT, the match has always been against the start string, taken literally, but perhaps I'm misunderstanding your comment.

In the catalog file I configure the resolver with, I have:
`<rewriteURI rewritePrefix="/path/to/a/directory" uriStartString="http://docbook.sourceforge.net/release/xsl-ns/current/"/>`. In version 1 this worked without the ending slash in the uriStartString; in version 3 unless the ending slash is there remainder of the URI is interpreted as absolute, leading to unresolvables like file:///html/chunk.xsl. Current behavior is likely correct, and I added the missing slashes in my catalog; it's just that version 1 seems to have been a bit more liberal in what it accepted :)

**Comment by dubinsky** [comment]
> And the `ValidatingXmlLoader` is only concerned with validating the catalog XML file, it has no impact on URI resolution. Can you clarify the issue you're describing?

When I add  `configuration.setFeature[java.lang.String](org.xmlresolver.ResolverFeature.CATALOG_LOADER_CLASS, "org.xmlresolver.loaders.ValidatingXmlLoader")` to my code, it slows down by around 50 seconds (from 3 to 53), and I suspect that the time is spent doing network requests...

**Comment by ndw** [comment]
With respect to `uriStartString`, I think you're correct. Without the trailing slash, the remainder of the URI begins with a "/" and that's going to resolve differently against the base URI.

**Comment by ndw** [comment]
I don't see how the validating loader could cause network requests. It explicitly points to a catalog that's in the jar file and the RNG file is also in the jar file. Can you share the catalog that takes 50 to parse with the validating loader?

**Comment by dubinsky** [comment]
> I don't see how the validating loader could cause network requests. It explicitly points to a catalog that's in the jar file and the RNG file is also in the jar file. Can you share the catalog that takes 50 to parse with the validating loader?

I'll get back to you on issue 4: I need to trace the execution a bit.

Some details on issue 3 - failure to resolve `file:` URIs with cache disabled:

With the cache enabled:
- `Resolver.resolve()` gets called with `href` "chunk.xsl"  and `baseURI` "file:/.../html/chunkfast.xsl";
- `CatalogResolver.resolveURI()` is called with the same parameters;
- at the `cache.cacheURI(absolute)` [call](https://github.com/xmlresolver/xmlresolver/blob/main/src/main/java/org/xmlresolver/CatalogResolver.java#L114) `absolute` is (correctly) "file:/.../html/chunk.xsl";
- `cache.cacheURI()` call returns `true`;
- `CatalogResolver.resource()` call returns a resource wrapping "file:/.../html/chunk.xsl";
- this resource becomes the overall result of the call to `Resolver.resolve()`.

With the cache disabled:
- `cache.cacheURI(absolute)` is called with the (correct) value of the `absolute` parameter ("file:/.../html/chunk.xsl");
- since `ResourceCache.cacheDir`=`null`, it [returns](https://github.com/xmlresolver/xmlresolver/blob/main/src/main/java/org/xmlresolver/cache/ResourceCache.java#L464) `false`;
- `CatalogResolver.resolveURI()` returns `null`;
- `Resolver.resolve()` returns `null`, discarding result of the resolution to "file:/.../html/chunk.xsl"...

**Comment by dubinsky** [comment]
My feeling is that the URIs that are excluded from the cache should be processed and (corresponding resources) returned **even if the cache is disabled**, but right now they are short-circuited to `null` in that scenario...

**Comment by dubinsky** [comment]
Concerning issue 4: when `ValidatingXmlLoader` is configured, every call to `Resolver.resolve()` causes an attempt (by 'org.relaxng:jing') to resolve "http://www.oasis-open.org/committees/entity/release/1.0/catalog.dtd". This resolution then fails ([again](#26)), causing re-retrieval of the file over the network at every call to resolver...

**Comment by ndw** [comment]
Concerning issue 3, I'm confused by one part of your analysis.

> * at the cache.cacheURI(absolute) call absolute is (correctly) "file:/.../html/chunk.xsl";
> * cache.cacheURI() call returns true;

The `cacheURI` method should never return `true` for a `file:` URI. It is *possible* to configure the resolver so it will, but you have to work at it. The default is never to cache `file:` URIs because the resolver returns the cached versions instead of the local versions. That's almost certainly an error.

The question of what the resolver should do in the case where `cacheURI` returns false is interesting. My experience with Java[*] parsers is that the resolver is expected to return `null` unless it is doing something more than just resolving the URI. In other words, if it fails to find some kind of redirection, it should return `null`.

Here, for example, is how the the return value from the SAX `resolveEntity` method in the `EntityResolver2` interface is described:

> An InputSource object describing the new input source to be used by the parser. Returning null directs the parser to resolve the system ID against the base URI and open a connection to resulting URI.

I think it's an open question whether the resolver would be *wrong* to return the resolved URI, but since the parser is going to turn around and arrive at exactly the same result, I've always returned null.

[*] It's interesting to note that the C# resolver API method is `getEntity()` and it is *always* expected to return the entity, even if it just opens up what it was passed.

**Comment by ndw** [comment]
I think points 3 and 4 are resolved. Regarding point 7, thank you. I really haven't got my head around the various logging frameworks and what things are actually dependent on what and how. I will try to clean this up, but specific suggestions are most welcome.

**Comment by dubinsky** [comment]
> ... if it fails to find some kind of redirection, it should return null.

Thank you for explaining this.

**Comment by dubinsky** [comment]
> I really haven't got my head around the various logging frameworks and what things are actually dependent on what and how. I will try to clean this up, but specific suggestions are most welcome.

I'd:
- move [slf4j-simple](https://github.com/xmlresolver/xmlresolver/blob/main/build.gradle#L52) from `testImplementation` to `testRuntimeOnly`;
- remove `runtimeOnly` [block](https://github.com/xmlresolver/xmlresolver/blob/main/build.gradle#L44): all the dependencies in it are also in `implementation`;
- move [slf4j-api](https://github.com/xmlresolver/xmlresolver/blob/main/build.gradle#L37) from `implementation` to `compileOnly`;

**Comment by dubinsky** [comment]
> The cacheURI method should never return true for a file: URI.

In the currently released version it does (when the cache is enabled); do I understand correctly that your recent changes fix that?

**Comment by ndw** [comment]
Yes, but I'm still surprised it's happening. I fixed a case that (I thought) could only arise because my cache was shared between my Java implementation and my nascent C# implementation.

Can you show me the .../cache/control.xml file that you're system is using?

**Comment by dubinsky** [comment]
> Can you show me the .../cache/control.xml file that you're system is using?

I do not have one...

**Comment by ndw** [comment]
If I understood correctly, you said that `cacheURI()` returns true when you have the cache enabled. If you have the cache enabled, there should be a file called `control.xml` in the cache directory. Is it not present?

**Comment by dubinsky** [comment]
No, it is not present. Whatever was supposed to write it didn't.

**Comment by ndw** [comment]
That is ... really weird. Okay, I'll take another look. If you can pull the latest and give it a try, I'd be curious if the problem is fixed.

**Comment by dubinsky** [comment]
Is the latest 3.0.2-SNAPSHOT published in some Maven/Gradle package repository?

**Comment by ndw** [comment]
Sorry. Just did that now, should be here:

```
    maven { url "https://oss.sonatype.org/content/repositories/snapshots" }
```

within a few minutes.

**Comment by dubinsky** [comment]
>  If you can pull the latest and give it a try, I'd be curious if the problem is fixed.

It is not. My thoughts are up top under issue 2.

**Comment by ndw** [comment]
Right. I think that's fixed by #49 and I've re-published the snapshot so you should be able to test it fairly easily. Thank you for all the excellent feedback and for your patience.

**Comment by dubinsky** [comment]
Yay! With the latest snapshot, everything works as it should!

I would still clean up dependencies in the `build.gradle` a bit:
- remove the `runtimeOnly` block - everything in there is already listed in the `implementation` block; this does not change anything for the users of the library - just the aesthetics of the Gradle build file itself;
- move 'org.slf4j:slf4j-api' from the `implementation` block to the `compileOnly` block to better account for the slf4j implementations that embed the API classes.

**Comment by ndw** [comment]
Yes, thanks. I will try to follow that advice and make another commit this evening (GMT+01:00). I think there's enough progress here to justify an actual 3.0.2 release, but I won't do that before tomorrow.

**Comment by dubinsky** [comment]
Thank you! Looking forward to the release.

**Comment by ndw** [comment]
I believe all of these issues are resolved in 3.0.2 released this morning. Thank you!

**Comment by dubinsky** [comment]
> ... 3.0.2 released this morning.

Thank you!
